### PR TITLE
Refactor qapply to use SlidingTransform and multiple dispatch

### LINKED ISSUE #27526: TensorProduct is not associative in all circumstances
Tensor products are [associative](https://en.wikipedia.org/wiki/Tensor_product#Properties). This property _is_ reflected in sympy's `TensorProduct`, but there are situations where this breaks:

```python
from sympy import expand
from sympy.physics.quantum import Ket, qapply, TensorProduct, Dagger

k0 = Ket("0")
k1 = Ket("1")

#define some operators (outer products)
I = k0*Dagger(k0) + k1*Dagger(k1)
k = k0*Dagger(k0)

# Example 1: These are the same. So far so good.
res1 = TensorProduct(I, TensorProduct(k, k))
res2 = TensorProduct(I, k, k)
print(qapply(expand(res1, tensorproduct=True)))
print(qapply(expand(res2, tensorproduct=True)))

# Example 2: These are NOT the same, but should be.
res1 = TensorProduct(I, TensorProduct(k, k))*TensorProduct(TensorProduct(k, k), I)
res2 = TensorProduct(I, k, k)*TensorProduct(k, k, I)
print(qapply(expand(res1, tensorproduct=True)))
print(qapply(expand(res2, tensorproduct=True)))
```

I know that the second example could be "fixed" by simply using `qapply` before the problematic multiplication (since this would turn the second example into the first example), but I came across this problem during a longer computation where such a "fix" does not work. 

Sympy verion: Main as of today https://github.com/sympy/sympy/commit/f07466ae38d6f7985c4e9eec2c7dfff43fec3cf7

### PR DESCRIPTION
**Pull request by ellisonbg**
#### References to other Issues or PRs

Related to #28022
Replacement for #24349
Fixes #27526, #27213

#### Brief description of what is fixed or changed

This PR introduces a major refactoring of the quantum physics package's operator application system. The key changes include:

1. New SlidingTransform Algorithm: Extracted and generalized the transformation algorithm used in quantum expression canonicalization into a reusable slidingtransform.py module. This algorithm provides a systematic way to apply transformations to multiplication expressions using a sliding window approach, which is particularly useful for non-commutative quantum operators.
2. Complete qapply Refactoring: Extensively refactored the qapply function to use the new SlidingTransform algorithm, improving its ability to handle complex quantum expressions involving operators, states, tensor products, and other quantum constructs.
3. Sum/Integral handlings: Sums and integrals are now properly handled in qapply.
4. Multiple-dispatch: The refactor uses multipledispatch to provide an extensible and clean way of defining how qapply works on different types of expressions.
5. No expand: qapply no longer calls `.expand()` as it works, resulting in more efficient calculations and results that are more similar to how humans do these types of calculations.
6. Less recursion: While recursion is still used in the algorithms, it is used in fewer places, resulting in a more efficient approach.
7. Enhanced Test Coverage: Added comprehensive tests for both the new SlidingTransform functionality and the refactored qapply system, including edge cases and complex quantum expressions.
8. Bug Fixes and Improvements: Fixed various issues in quantum operator applications, particularly around Grover operators, spin systems, and TensorProduct handling.

The refactoring is mostly backward compatibility while providing a more robust and extensible foundation for quantum symbolic computation.

#### Other comments

The SlidingTransform algorithm operates in two phases:

- Unary Phase: Processes individual factors in multiplications
- Binary Phase: Applies binary transformations to adjacent pairs using a sliding window

This approach allows for better handling of non-commutative quantum operators and provides a foundation for future quantum computation enhancements.

#### Release Notes

- physics.quantum
  - Introduced new SlidingTransform algorithm for systematic processing of quantum multiplication expressions
  - Major refactoring of qapply function for improved operator application handling
  - Enhanced support for complex quantum expressions involving tensor products and non-commutative operators
  - Full handling of sums and integrals in qapply
  - The qapply function no longer calls `.expand()` while it works so some of the results may have a slightly different form.
  - Fixed various bugs in Grover operators, spin systems, and quantum state applications
  - Improved test coverage for quantum operator applications and edge cases

### CODE REVIEW

**Comment by ellisonbg** [comment]
@oscarbenjamin @akhmerov this PR has the qapply refactor we discussed in the previous iteration of the PR (#28023). The biggest changes since you last looked at it are comprehensive testing. This is ready for final review.

Also @Matt-Ord may find this of interest.

File: sympy/physics/quantum/slidingtransform.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+    @property
+    def binary(self):
+        return self._binary
+
+    @property
+    def from_args(self):
+        return self._from_args
+
+    def __call__(self, expr, **options):
```
Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: expr should be type annotated as mul if it only accepts mul

Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: Return type should be annotated as sp.Expr

Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: Would sliding _transform be a better module name here?

File: sympy/physics/quantum/slidingtransform.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                    if last.is_commutative:
+                        c_parts.append(last)
+                    else:
+                        output.insert(0, input[-1])
+
+            input = output
+            output = []
+
+        nc_parts = tuple(input)
+        c_parts = tuple(c_parts)
```
Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: If both transforms were None is this correct?

File: sympy/physics/quantum/slidingtransform.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    l1, l2 = tee((condition(item), item) for item in seq)
+    return tuple(i for p, i in l1 if p), tuple(i for p, i in l2 if not p)
+
+
+def _split_cnc(seq):
+    c, nc = _split_on_condition(seq, lambda x: x.is_commutative)
+    return c, nc
+
+
+class SlidingTransform(object):
```
Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: Is there a massive perf impact to having a SlidingUnaryTransform followed by a SlidingBinaryTransform? This would make the logic in __call__ much simpler (as you wouldn't need to support None).

File: sympy/physics/quantum/slidingtransform.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+    def __init__(self, unary=None, binary=None, reverse=False, from_args=True):
```
Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: Type annotations needed here

File: sympy/physics/quantum/qapply.py [hunk 5]
```diff
@@ -1,232 +1,400 @@
...
+    return sympify(expr)
+
+
+#-----------------------------------------------------------------------------
+# qapply
+#-----------------------------------------------------------------------------
 
 def qapply(e, **options):
-    """Apply operators to states in a quantum expression.
+    """Apply quantum operators to quantum states in expressions.
```
Comment (Matt-Ord) [inline sympy/physics/quantum/qapply.py:201]: Needs type annotations , particularly for options

File: sympy/physics/quantum/qapply.py [hunk 6]
```diff
@@ -1,232 +1,400 @@
...
+    return None
+
 
+@_qapply_unary.register((int, float, complex))
+def _qapply_unary_number(expr, **options):
+    """Convert numbers to SymPy expressions."""
+    return sympify(expr)
+
+
+#-----------------------------------------------------------------------------
```
Comment (Matt-Ord) [inline sympy/physics/quantum/qapply.py:196]: What information does this comment provide?

File: sympy/physics/quantum/qapply.py [hunk 7]
```diff
@@ -1,232 +1,400 @@
...
+                result.extend(item.args)
+            else:
+                result.append(item)
+        return tuple(result)
+
+    return g
+
+
+#-----------------------------------------------------------------------------
+# _qapply_unary
```
Comment (Matt-Ord) [inline sympy/physics/quantum/qapply.py:105]: Remove comment?

**Review by Matt-Ord — CHANGES_REQUESTED** [review]
Left a few comments but I haven't yet looked at the tests. Liked the use of Dispatcher - it makes it a lot cleaner to read

File: sympy/physics/quantum/slidingtransform.py (hunk 3 above)
Comment (Matt-Ord) [inline sympy/physics/quantum/slidingtransform.py]: Does it need to inherit from object?

File: sympy/physics/quantum/qapply.py (hunk 7 above)
Comment (ellisonbg) [reply sympy/physics/quantum/qapply.py:105]: I am using these to organize the file. There is a lot going on, so it was helpful to have some prominent visual indicators of the different section.

File: sympy/physics/quantum/qapply.py (hunk 6 above)
Comment (ellisonbg) [reply sympy/physics/quantum/qapply.py:196]: Helps to organize the file for developers.

File: sympy/physics/quantum/slidingtransform.py (hunk 3 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: In Python 3 object is optional. I can remove, but the two are equivalent.

Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: There is a performance hit to having to walk through Mul args twice, but from an algorithm perspective, it is really needed, regardless of where it is implemented. I do think there is room for performance improvements in all of this, but I think the opportunities are one level deeper in the underlying APIs of sympy that I am using.

File: sympy/physics/quantum/slidingtransform.py (hunk 1 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: In the quantum package, we have been using file names without the underscores.

File: sympy/physics/quantum/slidingtransform.py (hunk 2 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Good question, let me right a test for that base.

File: sympy/physics/quantum/slidingtransform.py (hunk 3 above)
Comment (Matt-Ord) [reply sympy/physics/quantum/slidingtransform.py]: Yeah removing would be best - reading the code made me think it was inheriting from some other class called object

Comment (Matt-Ord) [reply sympy/physics/quantum/slidingtransform.py]: I would suggest splitting the class in two in this case - we could always add back a SlidingTransform that does both together if those optimisations get implemented at a later date.

File: sympy/physics/quantum/slidingtransform.py (hunk 1 above)
Comment (Matt-Ord) [reply sympy/physics/quantum/slidingtransform.py]: We might as well give things the correct name going forward though (or just make the module private if this is controversial)

File: sympy/physics/quantum/qapply.py (hunk 6 above)
Comment (Matt-Ord) [reply sympy/physics/quantum/qapply.py:196]: I think I would move it to a new file if this is really nessesary, but this isn't a hill I'd die on

File: sympy/physics/quantum/qapply.py (hunk 5 above)
Comment (ellisonbg) [reply sympy/physics/quantum/qapply.py:201]: Pushing new commits with type annotations.

File: sympy/physics/quantum/slidingtransform.py (hunk 3 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Fixed

File: sympy/physics/quantum/slidingtransform.py (hunk 4 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Fixed

File: sympy/physics/quantum/slidingtransform.py (hunk 3 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Because of how I use SlidingTransform in the Mul postprocessors in transform.py, I think it is important to keep it in a single class that does a single trip through the postprocessor logic. Basically, I am fine having the combined class to avoid multiple postprocessors.

File: sympy/physics/quantum/slidingtransform.py (hunk 2 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Looks like the behavior is correct - I have added a test for this case.

**Comment by ellisonbg** [comment]
@Matt-Ord OK I have addressed all of your comments - CI is running again...

**Comment by Matt-Ord** [comment]
> @Matt-Ord OK I have addressed all of your comments - CI is running again...

Brilliant - I won't have time to get to it today, but tomorrow morning I'll do a final review

**Comment by Matt-Ord** [comment]
@ellisonbg I've made a small change to the types [in this pr](https://github.com/ellisonbg/sympy/pull/2) so that qapply mul also gets kwargs type checked, other than that it looks good to me!

**Review by Matt-Ord — APPROVED** [review]

File: sympy/physics/quantum/slidingtransform.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                    if last.is_commutative:
+                        c_parts.append(last)
+                    else:
+                        output.insert(0, input[-1])
+
+            input = output
+            output = []
+
+        nc_parts = tuple(input)
+        c_parts = tuple(c_parts) # type: ignore[assignment]
```
Comment (oscarbenjamin) [inline sympy/physics/quantum/slidingtransform.py]: Rather than `type: ignore` it is better to do this kind of thing:
```diff
diff --git a/sympy/physics/quantum/slidingtransform.py b/sympy/physics/quantum/slidingtransform.py
index ae96d809a4..505b0e63db 100644
--- a/sympy/physics/quantum/slidingtransform.py
+++ b/sympy/physics/quantum/slidingtransform.py
@@ -26,7 +26,7 @@
 """
 
 from __future__ import annotations
-from typing import Any, Callable, Generic, Iterable, ParamSpec, TypeVar, Protocol
+from typing import Any, Callable, Generic, Iterable, ParamSpec, TypeVar, Protocol, Sequence
 
 from itertools import tee
 
@@ -163,8 +163,8 @@ def __call__(self, expr: Mul, *args: P.args, **kwargs: P.kwargs) -> Expr:
             raise TypeError('The SlidingTransform only works on Mul instances.')
 
         input = list(expr.args)
-        output = []
-        c_parts = []
+        output: list[Expr] = []
+        c_parts: list[Expr] = []
         if self._unary is not None:
             while len(input) > 0:
                 next = input.pop(0)
@@ -271,7 +271,6 @@ def __call__(self, expr: Mul, *args: P.args, **kwargs: P.kwargs) -> Expr:
             output = []
 
         nc_parts = tuple(input)
-        c_parts = tuple(c_parts) # type: ignore[assignment]
 
         # In the logic below, we go out of our way to avoid triggering the post-processor logic
         # unless it is absolutely needed by using Mul._from_args and always validating when we
@@ -284,8 +283,10 @@ def __call__(self, expr: Mul, *args: P.args, **kwargs: P.kwargs) -> Expr:
                 c_result_args = m.args
             else:
                 c_result_args = () if m == S.One else (m,)
+        elif c_parts and c_parts[0] == S.One:
+            c_result_args = ()
         else:
-            c_result_args = () if (c_parts and c_parts[0] == S.One) else c_parts  # type: ignore[assignment]
+            c_result_args = tuple(c_parts)
 
         # Handle the non-commutative part
         if len(nc_parts) > 1:
```
Then the type checkers actually understand the types in the body of the function (right up to `Mul._from_args` at the end).

File: sympy/physics/quantum/slidingtransform.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+
+def _split_on_condition(seq: Iterable[T], condition: Callable[[T], bool]) -> tuple[tuple[T, ...], tuple[T, ...]]:
+    l1, l2 = tee((condition(item), item) for item in seq)
+    return tuple(i for p, i in l1 if p), tuple(i for p, i in l2 if not p)
+
+
+def _split_cnc(seq: Iterable[Any]) -> tuple[tuple[Any, ...], tuple[Any, ...]]:
+    c, nc = _split_on_condition(seq, lambda x: x.is_commutative)
+    return c, nc
```
Comment (oscarbenjamin) [inline sympy/physics/quantum/slidingtransform.py:55]: Why is `Any` used here?

I would have thought that this should be
```python
def _split_on_condition(seq: Iterable[T], condition: Callable[[T], bool | None]) -> tuple[tuple[T, ...], tuple[T, ...]]:
    l1, l2 = tee((condition(item), item) for item in seq)
    return tuple(i for p, i in l1 if p), tuple(i for p, i in l2 if not p)


def _split_cnc(seq: Iterable[Expr]) -> tuple[tuple[Expr, ...], tuple[Expr, ...]]:
    c, nc = _split_on_condition(seq, lambda x: x.is_commutative)
    return c, nc
```

Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py:55]: Good catch, pushing a commit that fixes this.

File: sympy/physics/quantum/slidingtransform.py (hunk 8 above)
Comment (ellisonbg) [reply sympy/physics/quantum/slidingtransform.py]: Just pushed a commit to fix this.

**Comment by ellisonbg** [comment]
Review comments addressed, tests running again
